sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala

```diff
@@ -50,6 +55,66 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
       .createOrReplaceTempView("orc_temp_table")
   }
 
+  protected def testBloomFilterCreation(bloomFilterKind: Kind) {
+    val tableName = "bloomFilter"
+
+    withTempDir { dir =>
+      withTable(tableName) {
+        val sqlStatement = orcImp match {
+          case "native" =>
+            s"""
+               |CREATE TABLE $tableName (a INT, b STRING)
+               |USING ORC
+               |OPTIONS (
+               |  path '${dir.toURI}',
+               |  orc.bloom.filter.columns '*',
+               |  orc.bloom.filter.fpp 0.1
+               |)
+            """.stripMargin
+          case "hive" =>
+            s"""
+               |CREATE TABLE $tableName (a INT, b STRING)
+               |STORED AS ORC
+               |LOCATION '${dir.toURI}'
+               |TBLPROPERTIES (
+               |  orc.bloom.filter.columns='*',
+               |  orc.bloom.filter.fpp=0.1
+               |)
+            """.stripMargin
+          case impl =>
+            throw new UnsupportedOperationException(s"Unknown ORC implementation: $impl")
+        }
+
+        sql(sqlStatement)
+        sql(s"INSERT INTO $tableName VALUES (1, 'str')")
+
+        val partFiles = dir.listFiles()
+          .filter(f => f.isFile && !f.getName.startsWith(".") && !f.getName.startsWith("_"))
+        assert(partFiles.length === 1)
+
+        val orcFilePath = new Path(partFiles.head.getAbsolutePath)
+        val readerOptions = OrcFile.readerOptions(new Configuration())
+        val reader = OrcFile.createReader(orcFilePath, readerOptions)
+        var recordReader: RecordReaderImpl = null
+        try {
+          recordReader = reader.rows.asInstanceOf[RecordReaderImpl]
+
+          // BloomFilter array is created for all types; `struct`, int (`a`), string (`b`)
+          val sargColumns = Array(true, true, true)
+          val orcIndex = recordReader.readRowIndex(0, null, sargColumns)
+
+          // Check the types and counts of bloom filters
+          assert(orcIndex.getBloomFilterKinds.forall(_ === bloomFilterKind))
```

Comment:
how can we extend it in the future? How can we change the bloom filter kind via the CREATE TABLE statement?

Reply:
Do you mean how we extend this test case? If so, I think it's fine since what we need to test within Spark is the specified bloom filter works or not. It's rather none of all so one test case should be okay.

Reply:
Thank you for review, @cloud-fan and @HyukjinKwon .

It's controlled by , `orc.bloom.filter.write.version`.
- `utf8` (default): Include only `BLOOM_FILTER_UTF8`.
- `original`: Include both `BLOOM_FILTER` and `BLOOM_FILTER_UTF8`.

As you see, `original` will populate both filters to support old and new reader simultaneously.

And, one additional related configuration is `orc.bloom.filter.ignore.non-utf8` (default: false) which enforces to ignore old BLOOM_FILTER explicitly in case of `true`.

Reply:
```scala
scala> spark.version
res1: String = 2.3.1
scala> sql("set spark.sql.orc.impl=native")
scala> sql("create table bf using orc options(orc.bloom.filter.columns '*') as select 1 a")
scala> sql("create table bf2 using orc options(orc.bloom.filter.columns '*', orc.bloom.filter.write.version 'original') as select 1 a")
```
```scala
$ orc-tools meta spark-warehouse/bf | grep BLOOM_FILTER
    Stream: column 0 section BLOOM_FILTER_UTF8 start: 14 length 381
    Stream: column 1 section BLOOM_FILTER_UTF8 start: 419 length 402

$ orc-tools meta spark-warehouse/bf2 | grep BLOOM_FILTER
    Stream: column 0 section BLOOM_FILTER start: 14 length 426
    Stream: column 0 section BLOOM_FILTER_UTF8 start: 440 length 381
    Stream: column 1 section BLOOM_FILTER start: 845 length 456
    Stream: column 1 section BLOOM_FILTER_UTF8 start: 1301 length 402
```

Reply:
It seems the test here  is creating orc with bloom filter using spark with options, and read it back through native ORC reader. Is there a plan to add an optimizer rule in Spark to show the functionality of this in the physical plan like predicate pushdown in parquet?

Reply:
Thank you for review, @dbtsai . Definitely, it would be great if we show some information.
BTW, could you elaborate more about `in the physical plan like predicate pushdown in parquet` specifically?

Reply:
Something like

```
== Physical Plan ==
*(1) Project [_1#3]
+- *(1) Filter (isnotnull(_1#3) && (_1#3._1 = true))
   +- *(1) FileScan parquet [_1#3] Batched: false, Format: Orc, 
  PushedFilters: [IsNotNull(_1), EqualTo(_1._1,true)]
  BloomFilters: [some information]
```

Thanks.

Reply:
Got it. Thanks.

Reply:
FWIW, there might be one more thing consider in Parquet too - block level filter and record level filter.

and if pushed filters are actually handled within each source properly or not (fileformat specifically).